File: domain/conversation/conversation.py
"""Conversation aggregate - manages the session transcript and emotion events."""
from datetime import datetime, timezone
from typing import List

from .message import Message, MessageRole


class Conversation:
    """Aggregate root for a single conversation session.

    Holds the ordered list of messages and any observed emotion events
    that occurred during the session.
    """

    def __init__(self) -> None:
        self.messages: List[Message] = []
        self.emotion_events: List[str] = []
        self.started_at: datetime = datetime.now(timezone.utc)

    def add_message(self, role: str, text: str) -> Message:
        """Append a message and return the new Message entity."""
        message = Message(role=role, text=text)
        if not message.is_empty():
            self.messages.append(message)
        return message

    def add_emotion_event(self, emotion: str) -> None:
        """Record an observed user emotion during the session."""
        if emotion:
            self.emotion_events.append(emotion)

    def get_transcript(self) -> List[dict]:
        """Return transcript as a list of plain dicts for serialization."""
        return [m.to_dict() for m in self.messages]

    def is_empty(self) -> bool:
        return len(self.messages) == 0

    @property
    def message_count(self) -> int:
        return len(self.messages)

File: src/domain/conversation/Conversation.js
/**
 * Conversation - aggregate root for a single chat session.
 *
 * Holds the ordered list of messages and any user emotion events observed
 * during the session (used when saving memory at the end).
 */

import { Message, MessageRole } from "./Message.js";

export class Conversation {
  constructor() {
    /** @type {Message[]} */
    this.messages = [];
    /** @type {string[]} */
    this.emotionEvents = [];
    this.startedAt = new Date();
  }

  /**
   * Append a message if non-empty, returning the Message entity.
   * @param {string} role
   * @param {string} text
   * @returns {Message}
   */
  addMessage(role, text) {
    const message = new Message({ role, text });
    if (!message.isEmpty()) {
      this.messages.push(message);
    }
    return message;
  }

  /**
   * Record an observed user emotion event for the session summary.
   * @param {string} emotion
   */
  addEmotionEvent(emotion) {
    if (emotion) {
      this.emotionEvents.push(emotion);
    }
  }

  /** Return transcript as plain dicts for serialization. */
  getTranscript() {
    return this.messages.map((m) => m.toDict());
  }

  isEmpty() {
    return this.messages.length === 0;
  }

  get messageCount() {
    return this.messages.length;
  }
}
